File: onebusaway-android/src/main/java/org/onebusaway/android/nav/FeedbackReceiver.java
/*
 * Copyright (C) 2019 University of South Florida
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.nav;

import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.analytics.FirebaseAnalytics;

import org.apache.commons.io.FileUtils;
import org.onebusaway.android.BuildConfig;
import org.onebusaway.android.R;
import org.onebusaway.android.app.Application;
import org.onebusaway.android.io.ObaAnalytics;
import org.onebusaway.android.ui.FeedbackActivity;
import org.onebusaway.android.ui.HomeActivity;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import androidx.core.app.NotificationCompat;
import androidx.core.app.RemoteInput;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import static android.text.TextUtils.isEmpty;
import static org.onebusaway.android.nav.NavigationService.LOG_DIRECTORY;

public class FeedbackReceiver extends BroadcastReceiver {
    public static final String TAG = "FeedbackReceiver";
    public static final String ACTION_REPLY = BuildConfig.APPLICATION_ID + ".action.REPLY";
    public static final String ACTION_DISMISS_FEEDBACK = BuildConfig.APPLICATION_ID + ".action.DISMISS_FEEDBACK";
    public static final String TRIP_ID = ".TRIP_ID";
    public static final String NOTIFICATION_ID = ".NOTIFICATION_ID";
    public static final String RESPONSE = ".RESPONSE";
    public static final String LOG_FILE = ".LOG_FILE";
    public static final String SHOW_SNACKBAR = "show_snackbar";

    public static final int FEEDBACK_NO = 1;
    public static final int FEEDBACK_YES = 2;

    private FirebaseAnalytics mFirebaseAnalytics;

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "onReceive");

        mFirebaseAnalytics = FirebaseAnalytics.getInstance(context);

        int notifyId = intent.getIntExtra(NOTIFICATION_ID,
                NavigationServiceProvider.NOTIFICATION_ID + 1);
        //int actionNum = intent.getIntExtra(ACTION_NUM, 0);
        String action = intent.getAction();

        if (ACTION_DISMISS_FEEDBACK.equals(action)) {
            Log.d(TAG, "Dismiss intent - showing Snackbar");
            showFeedbackDismissedSnackbar(context, intent);
        } else if (ACTION_REPLY.equals(action)) {
            Log.d(TAG, "Capturing user feedback from notification");
            captureFeedback(context, intent, notifyId);
        }
    }

    private void captureFeedback(Context context, Intent intent, int notifyId) {
        int response = intent.getIntExtra(RESPONSE, 0);
        String logFile = intent.getExtras().getString(LOG_FILE);
        Bundle remoteInput = RemoteInput.getResultsFromIntent(intent);
        String userResponse = null;

        if (remoteInput != null) {
            Log.d(TAG, "checked remoteInput ");
            String feedback = remoteInput.getCharSequence(
                    NavigationService.KEY_TEXT_REPLY).toString();

            Log.d(TAG, "Feedback captured");

            NotificationCompat.Builder repliedNotification =
                    new NotificationCompat.Builder(context
                            , Application.CHANNEL_DESTINATION_ALERT_ID)
                            .setSmallIcon(R.drawable.ic_stat_notification)
                            .setContentTitle(Application.get().getResources()
                                    .getString(R.string.feedback_notify_title))
                            .setContentText(Application.get().getResources()
                                    .getString(R.string.feedback_notify_confirmation));

            Log.d(TAG, "Thank you!");
            repliedNotification.setOngoing(false);

            NotificationManager mNotificationManager = (NotificationManager)
                    Application.get().getSystemService(Context.NOTIFICATION_SERVICE);
            mNotificationManager.notify(notifyId, repliedNotification.build());

            if (response == FEEDBACK_YES) {
                userResponse = Application.get().getString(R.string.analytics_label_destination_reminder_yes);
            } else {
                userResponse = Application.get().getString(R.string.analytics_label_destination_reminder_no);
            }

            Log.d(TAG, "cancelling notification");
            mNotificationManager.cancel(notifyId);

            Boolean pref = Application.getPrefs().getBoolean(Application.get().getResources()
                    .getString(R.string.preferences_key_user_share_destination_logs), true);

            if (pref) {
                Log.d(TAG, "True");
                moveLog(feedback, userResponse, logFile);
            } else {
                Log.d(TAG, "False");
                deleteLog(logFile);
                logFeedback(feedback, userResponse);
            }
        }
    }

    private void moveLog(String feedback, String userResponse, String logFile) {
        try {
            File lFile = new File(logFile);
            FileUtils.write(lFile, System.getProperty("line.separator") + "User Feedback - " + feedback, true);
            Log.d(TAG, "Feedback appended");

            File destFolder = new File(Application.get().getApplicationContext().getFilesDir()
                    .getAbsolutePath() + File.separator + LOG_DIRECTORY + File.separator + userResponse);

            Log.d(TAG, "sourceLocation: " + lFile);
            Log.d(TAG, "targetLocation: " + destFolder);

            try {
                FileUtils.moveFileToDirectory(
                        FileUtils.getFile(lFile),
                        FileUtils.getFile(destFolder), true);
                Log.d(TAG, "Move file successful.");
            } catch (Exception e) {
                Log.d(TAG, "File move failed");
            }

            setupLogUploadTask();

        } catch (IOException e) {
            Log.e(TAG, "File write failed: " + e.toString());
        }

    }

    private void deleteLog(String logFile) {
        File lFile = new File(logFile);
        boolean deleted = lFile.delete();
        Log.v(TAG, "Log deleted " + deleted);
    }

    private void setupLogUploadTask() {
        PeriodicWorkRequest.Builder uploadLogsBuilder =
                new PeriodicWorkRequest.Builder(NavigationUploadWorker.class, 24,
                        TimeUnit.HOURS);

        // Create the actual work object:
        PeriodicWorkRequest uploadCheckWork = uploadLogsBuilder.build();

        // Then enqueue the recurring task:
        WorkManager.getInstance().enqueue(uploadCheckWork);
    }

    private void logFeedback(String feedbackText, String userResponse) {
        Boolean wasGoodReminder;
        if (userResponse.equals(Application.get().getString(R.string.analytics_label_destination_reminder_yes))) {
            wasGoodReminder = true;
        } else {
            wasGoodReminder = false;
        }
        ObaAnalytics.reportDestinationReminderFeedback(mFirebaseAnalytics, wasGoodReminder
                , ((!isEmpty(feedbackText)) ? feedbackText : null), null);
        Log.d(TAG, "User feedback logged to Firebase Analytics :: wasGoodReminder - "
                + wasGoodReminder + ", feedbackText - " + ((!isEmpty(feedbackText)) ? feedbackText : null));
    }

    /**
     * Shows a Snackbar when feedback notification is dismissed
     */
    private void showFeedbackDismissedSnackbar(Context context, Intent originalIntent) {
        // Get notification ID and optional extras
        int notifyId = originalIntent.getIntExtra(NOTIFICATION_ID, 
                NavigationServiceProvider.NOTIFICATION_ID + 1);
        String tripId = originalIntent.getStringExtra(TRIP_ID);
        String logFilePath = originalIntent.getStringExtra(LOG_FILE);
        
        // Launch HomeActivity to show the Snackbar
        Intent intent = new Intent(context, HomeActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        intent.putExtra(SHOW_SNACKBAR, true);
        intent.putExtra(NOTIFICATION_ID, notifyId);
        
        // Pass trip ID and log file if available
        if (tripId != null) {
            intent.putExtra(TRIP_ID, tripId);
        }
        if (logFilePath != null) {
            intent.putExtra(LOG_FILE, logFilePath);
        }
        
        context.startActivity(intent);
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/nav/FeedbackReceiver.kt
/*
 * Copyright (C) 2019-2026 University of South Florida and Open Transit Software Foundation
 * Licensed under the Apache License, Version 2.0
 */
package org.onebusaway.android.nav

import android.app.NotificationManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.RemoteInput
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit
import org.onebusaway.android.BuildConfig
import org.onebusaway.android.R
import org.onebusaway.android.app.di.AnalyticsEntryPoint
import org.onebusaway.android.notifications.NotificationChannels
import org.onebusaway.android.util.PreferenceUtils

class FeedbackReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val notificationId = intent.getIntExtra(NOTIFICATION_ID, NavigationService.NOTIFICATION_ID + 1)
        when (intent.action) {
            ACTION_DISMISS_FEEDBACK -> Log.d(TAG, "Feedback notification dismissed")
            ACTION_REPLY -> captureFeedback(context, intent, notificationId)
        }
    }

    private fun captureFeedback(context: Context, intent: Intent, notificationId: Int) {
        val feedback = RemoteInput.getResultsFromIntent(intent)
            ?.getCharSequence(NavigationService.KEY_TEXT_REPLY)
            ?.toString()
            ?: return
        val response = intent.getIntExtra(RESPONSE, 0)
        val responseLabel = context.getString(
            if (response == FEEDBACK_YES) {
                R.string.analytics_label_destination_reminder_yes
            } else {
                R.string.analytics_label_destination_reminder_no
            }
        )
        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.notify(
            notificationId,
            NotificationCompat.Builder(context, NotificationChannels.DESTINATION_ALERT_ID)
                .setSmallIcon(R.drawable.ic_bus)
                .setContentTitle(context.getString(R.string.feedback_notify_title))
                .setContentText(context.getString(R.string.feedback_notify_confirmation))
                .setOngoing(false)
                .build()
        )
        manager.cancel(notificationId)

        val logPaths = intent.getStringArrayListExtra(LOG_FILES)
            ?.filter(String::isNotBlank)
            ?.ifEmpty { null }
            ?: listOfNotNull(intent.getStringExtra(LOG_FILE))
        val share = PreferenceUtils.getBoolean(
            context.getString(R.string.preferences_key_user_share_destination_logs),
            true
        )
        if (share && logPaths.isNotEmpty()) {
            moveLogs(context, feedback, responseLabel, logPaths)
        } else {
            logPaths.forEach { File(it).delete() }
            AnalyticsEntryPoint.get(context).reportDestinationReminderFeedback(
                response == FEEDBACK_YES,
                feedback.ifEmpty { null },
                null
            )
        }
    }

    private fun moveLogs(context: Context, feedback: String, response: String, paths: List<String>) {
        try {
            val destinationDirectory = File(
                File(context.filesDir, NavigationService.LOG_DIRECTORY),
                response
            ).apply { mkdirs() }
            paths.forEach { path ->
                val source = File(path)
                source.appendText(System.lineSeparator() + "User Feedback - " + feedback)
                val destination = File(destinationDirectory, source.name)
                if (!source.renameTo(destination)) {
                    throw IOException("Failed to move $source to $destination")
                }
            }
            val uploadWork = PeriodicWorkRequest.Builder(
                NavigationUploadWorker::class.java,
                24,
                TimeUnit.HOURS
            ).build()
            WorkManager.getInstance(context).enqueueUniquePeriodicWork(
                NavigationUploadWorker.UNIQUE_WORK_NAME,
                ExistingPeriodicWorkPolicy.KEEP,
                uploadWork
            )
        } catch (error: IOException) {
            Log.e(TAG, "File write failed", error)
        }
    }

    companion object {
        const val TAG = "FeedbackReceiver"
        const val ACTION_REPLY = BuildConfig.APPLICATION_ID + ".action.REPLY"
        const val ACTION_DISMISS_FEEDBACK = BuildConfig.APPLICATION_ID + ".action.DISMISS_FEEDBACK"
        const val TRIP_ID = ".TRIP_ID"
        const val NOTIFICATION_ID = ".NOTIFICATION_ID"
        const val RESPONSE = ".RESPONSE"
        const val LOG_FILE = ".LOG_FILE"
        const val LOG_FILES = ".LOG_FILES"
        const val FEEDBACK_NO = 1
        const val FEEDBACK_YES = 2
    }
}
